sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBLoader.scala

```diff
@@ -48,7 +48,16 @@ object RocksDBLoader extends Logging {
 
   def loadLibrary(): Unit = synchronized {
     if (exception == null) {
-      loadLibraryThread.start()
+      // SPARK-39847: if a task thread is interrupted while blocking in this loadLibrary()
+      // call then a second task thread might start a loadLibrary() call while the first
```

Comment:
I'm curious why the second task thread is not blocked by `synchronized` at line 48, @JoshRosen ?

Reply:
The second task thread is blocked by `synchronized`. Quoted the explain from Josh with some highlight:

- ...
- If Task A is interrupted, an `InterruptedException` will be thrown and **it will exit the loadLibrary synchronized block**.
- At this point, **Task B enters the synchronized block** of its `loadLibrary() call and sees that `exception == null` because the `loadLibraryThread` started by the other task is still running, so Task B calls `loadLibraryThread.start()` and hits the thread state error because it tries to start an already-started thread.

Reply:
Oh, got it. I missed that part. Thanks, @xuanyuanking .